rfcs/0000-execute-builds-in-standard-containers.md

```diff
@@ -0,0 +1,132 @@
+
+## Actions to take
+
+1. Create a new binary called `build-waiter`. This will have 2 primary responsibilities/modes of operation:
+    1. Copy Mode - This mode will be executed in an init container to copy the binary into a mounted volume that will
+       be mounted to every other container in the pod
+    2. Wait Mode - This mode will wait for a particular file to exist (For this implementation it will be a file
+       named for the step we are waiting on inside the shared mount that the `build-waiter` binary is in) and once it
+       does, it will execute the provided command. Upon completion of the command execution it will write out a file
+       corresponding to the build step that was just completed. If the command fails, the error will be written to an
+       error file that is being watched by every other container. If the error file is created, all other containers
+       will exit with the same error code, forcing the pod to enter a failed state. There is also a configurable
+       timeout period that will specify how long the container is going to wait for th eprevious steps to complete
+       before exiting with an error. Note: For `prepare` we can provide an empty `wait step` so that it begins
+       executing immediately.
+
+   Example of possible `build-waiter` implementation is in the [spike][kpack-spike-build-waiter].
+
+2. Convert the existing init containers in `build_pod.go` to be regular containers and add an init container to copy
+   the `build-waiter` to a shared mount
+3. Using a similar [step modifier][step-modifier] approach , we can put this new behavior behind a feature gate and
+   only execute in regular containers if enabled by the user. We can do this at the kpack controller level or at the
+   image level via a spec change or annotation. In the event that the feature flag is not set, kpack will continue
+   to execute builds in init containers.
+4. Due to the fact that sidecars such as istio-proxy, do not exit when all other containers exit, we will need to add a
+   method for stopping the proxy once a build exits
+    1. To do this we can update the running pod's `spec.activeDeadlineSeconds` to 1 which will cause the pod to go into
+       a failed state and kill all containers after 1 second. This will happen in the case that during a build
+       reconcile, kpack detects that a pod is running, but completion container is not. (The completion container will
+       exit in the event that another build container fails)
+    2. This will kill the sidecar, but it will also leave our pod in a failed state with the message `DeadlineExceeded`,
+       leading the reconciler to believe
+       the build has failed. To circumvent this we will add a check that if a pod has failed with the
+       reason `DeadlineExceeded` AND the completion container has successfully completed, then the build is actually
+       successful.
```

Comment:
If I'm not mistaken, Tekton uses a mechanism that involves changing the image in a sidecar container to a no-op image: https://github.com/tektoncd/pipeline/blob/119f7bdf2dd73bbd2a6b74506c75e000c7535f49/pkg/pod/entrypoint.go#L238
Would this be an avenue worth exploring?

Reply:
Yes, this is proposed in the first alternative towards the bottom. The main reason we liked activeDeadlineSeconds is that it is a pod level spec change, so you don't have to iterate over ever sidecar. Additionally the no-op image still results in a failed pod because you can not modify the pre-hook lifecycle that istio is running so the no-op exits with command not found. Tekton gets around this during the reconcile by ignoring that failure similar to how I am proposing. The main difference is that it is very easy to catch the activeDeadlineSeconds pod failure int he reconcile of a build

Reply:
Here is some other context from the tekton slack https://tektoncd.slack.com/archives/CLCCEBUMU/p1660923060292999